source/agora/consensus/ValidatorSet.d

```diff
@@ -397,6 +397,61 @@ public class ValidatorSet
         return PreImageInfo.init;
     }
 
+    /***************************************************************************
+
+        Get validator's pre-image for the given block height from the
+        validator set
+
+        Params:
+            enroll_key = The key for the enrollment in which the pre-image is
+                contained.
+            height = the desired preimage block height. If it's older than
+                     the preimage's current block height, the preimage will
+                     be hashed until the older preimage is retrieved.
+
+        Returns:
+            the PreImageInfo of the enrolled key if it exists,
+            otherwise PreImageInfo.init
+
+    ***************************************************************************/
+
+    public PreImageInfo getPreimageAt (const ref Hash enroll_key,
```

Comment:
This function is kind of duplicate with the `getPreimage` function. So it would be good to merge this function with the original 'getPreimage`.

Reply:
I'm not sure I can merge them. The use-cases are different.

Take a look at this call:

https://github.com/bpfkorea/agora/blob/56cd093013a57cd6f37c0fd9f7caaf663d256bd6/source/agora/consensus/ValidatorSet.d#L414

If I add a `height` parameter to `getPreimage`, what do we pass as a height to the `getPreimage` call?

In this case we just want the latest preimage, with no condition on the enrollment height or the distance.

Reply:
> If I add a `height` parameter to `getPreimage`, what do we pass as a height to the `getPreimage` call?
> 
> In this case we just want the latest preimage with no condition on the enrollment height or the distance.

There is only one pre-image for the `enroll_key` in the table, so it can be just selected and you can calculate the proper pre-image with several hashing.

Reply:
Two examples:

- Validator submitted preimage for block height 20
- We just externalized block height 10
- We want to get the preimage at height 10. So we hash the existing preimage 10 times to get the preimage at height 10.

- The Validator wants to submit the preimage for block height 30
- We should not hash anything here, just validate the new preimage and replace the old one.

There are two separate use-cases here.

Reply:
I thought the function can process with a default parameter for Example 2. That could make the function complex.
Thank you for your explanation.

Reply:
It's not really about complexity. It's not possible to combine the two functions together.

In example 1 if the stored preimage is height 9 but we want a preimage for height 10 then the function should return PreImageInfo.init. So we cannot use `Height(ulong.max)` as the default value because it would return the preimage at height 9.

And if we used `Height(0)` as the default value then the function will unnecessarily hash the value when it's used in example 2.